### ISSUE TITLE
Thread safety guarantees.

### ISSUE THREAD

**Issue by bader** [KhronosGroup/SYCL-Docs#415]
From https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#chapter:sycl-programming-interface

> SYCL guarantees that all the member functions and special member functions of the SYCL classes described are thread safe.

Some functionality is exposed as free functions or function templates i.e. not member functions or special member functions. Does SYCL guarantee thread safety? If so, should we make the wording more clear/precise?

**Comment by gmlueck** [comment]
Yes, I think  that makes sense.  We could say something like:

> All functions defined in this specification are thread safe.

**Comment by Pennycook** [comment]
> Yes, I think that makes sense. We could say something like:
> 
> > All functions defined in this specification are thread safe.

I might be overthinking the meaning of "thread safe", but that seems quite strong.  I don't think the member functions of `accessor` are thread safe, for example, and I'm not sure about the behavior of `sycl::stream`.

What about:

> All functions defined in this specification are thread safe, unless otherwise specified.

Then we have the freedom to clarify other behavior, and won't forget to update this sentence if we do.

**Comment by gmlueck** [comment]
> Then we have the freedom to clarify other behavior, and won't forget to update this sentence if we do.

Yes, that makes sense.

We say the following about `stream`:

> The SYCL stream class is required to output the content of each stream, between flushes (up to workItemBufferSize), without mixing with content from the same stream in other work-items. There are no other output order guarantees between work-items or between streams. The stream flush operation therefore delimits the unit of output that is guaranteed to be displayed without mixing with other work-items, with respect to a single stream.

What is not thread safe with accessor?

**Comment by Pennycook** [comment]
> We say the following about `stream`:

Interesting.  I think that's stronger than the guarantees provided by `std::stream`, but deliberately so.

> What is not thread safe with accessor?

I'm still worried I'm overthinking this, but here's an example.

If two threads (i.e. work-items) write concurrently to the same location in an accessor, that's a race condition.  The accessor class provides an assignment operator function (as a member function) for zero-dimension accessors, and calling that function from multiple threads is unsafe.

I think this is subtly different to the multi-dimensional case.  There, the subscript operator returns a `T&` -- the member function can safely be called by multiple threads, and it's the developer's responsibility to avoid introducing data races when they use the return value.

**Comment by gmlueck** [comment]
Yes, I think you are right.  This is roughly equivalent to the STL containers, where each thread can write to a different object in the container, but multiple threads cannot safely write to the same object.
